### PR TITLE
Support using the reconnecting ldap client.

### PR DESCRIPTION
**Pull request by jdavisp3**
@lucaswiman @naegelyd 

Update Baya to support using the reconnecting client (`ldap.ldapobject.ReconnectLDAPObject`) from the `python-ldap` package via the `BAYA_USE_RECONNECTING_CLIENT` setting.

In testing Baya against Active Directory we have found that approximately 1% of LDAP queries will fail with `ldap.SERVER_DOWN`. The root cause is unclear, as is whether the problem is on the client or server side. In any case, the `ReconnectLDAPObject` is designed for backends where this happens:

> In case of server failure (ldap.SERVER_DOWN) the implementations
> of all synchronous operation methods (search_s() etc.) are doing
> an automatic reconnect and rebind and will retry the very same
> operation.

So this PR adds a `BAYA_USE_RECONNECTING_CLIENT` setting which, if set `True`, will use the reconnecting client in the `NestedLDAPGroupsBackend`. Because of the way the actual connection is made inside of `django_auth_ldap` this isn't as straightforward as it could be. We use a shim, `ReconnectingLDAP` that behaves like the `ldap` module except `ldap.initialize` uses the reconnecting ldap object.

There are unit tests around the shim. This was also tested with a script that issues 20K LDAP queries and recording any failures.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +90,15 @@ BAYA_LOGIN_URL = "/internal/login/url"
 
 Of course, change the values to match your actual setup.
 
+## LDAP client configuration
+
+By default, Baya uses `ldap.initialize` to make the LDAP connection. To use the
+`ReconnectLDAPObject` provided by the `ldap` package, configure this setting:
+
+```
+BAYA_USE_RECONNECTING_CLIENT
```
Comment (lucaswiman) [inline README.md]: `BAYA_USE_RECONNECTING_CLIENT = True`

File: baya/backend.py [hunk 2]
```diff
@@ -33,3 +48,31 @@ def get_all_permissions(self, user, obj=None):
...
+    attribute returns ReconnectLDAPObject instead of the original
+    initialize function (which just instantiates a SimpleLDAPObject).
+    """
+    retry_max = 6
+
+    def __init__(self, original_module):
+        self._original_module = original_module
+
+    def __getattr__(self, name):
+        return getattr(self._original_module, name)
```
Comment (lucaswiman) [inline baya/backend.py]: I know it's standard python, but I can never remember what the rarely-used subtleties of `__getattribute__`, `__getattr__` and `getattr` are. (I had to look it up.) Maybe add a comment reminding the reader that `__getattr__` is called when an attribute is not found in the instance `__dict__` or in the class MRO.

**Review by lucaswiman — APPROVED** [review]
Just a few clarifying suggestions. Code LGTM.

**Review by naegelyd — APPROVED** [review]
Could you add an entry to the [CHANGELOG](https://github.com/counsyl/baya/blob/master/CHANGELOG) for this new version?

LGTM2

File: README.md (hunk 1 above)
Comment (jdavisp3) [reply README.md]: Doh, thanks for catching that.

File: baya/backend.py (hunk 2 above)
Comment (jdavisp3) [reply baya/backend.py]: Good call, I had to look it up myself.

**Comment by jdavisp3** [comment]
Thanks very much @lucaswiman and @naegelyd !

**Pull request merged**